DiceBot 클래스를 추가

jwChung opened the pull request:
Dicebot 클래스를 추가합니다.

이 PR은 #5 이슈와 관련됩니다.
- [x] Create a skeleton of the dice class
- [x] Basic test code

jwChung committed:
불필요한 코드 삭제,

github를 테스트하기 위해 추가된 코드를 제거합니다.

jwChung commented:
머릿말 WIP(working in progress)는 의미로 사용했습니다. 아직 작업중이니 merge대상은 아닙니다. 구현을 끝내고PR을 올리는 것보다 좀 더 빨리 코드를 동료들에게 공유하기 위함입니다.

jwChung committed:
DiceBot.process 메소드 추가,

DiceBot 클래스를 생성하고, process 함수를 추가했습니다.

jwChung commented:
@funprog/core 리뷰 부탁합니다.

jwChung commented on src/main/scala/DiceBot.scala:
```diff
@@ -0,0 +1,3 @@
+object DiceBot {
+  def roll(input: String): Option[String] = None
```
오프라인 모임에서 `onReceive(input: String)` 라는 메소드로 얘기되었었는데 그냥 여기서 바로 `DiceBot.roll` 이름을 사용하면 어떨까요? 

`roll` 함수의 리턴 값은 `Option[String]`으로 유효하지 않은 input 값에 대해서는 `None` 을 반환하도록 하였습니다.

jwChung replied:
@myeesan 질문하나 드려도 될까요? 가령 사용자가 `/roll 3d6`이라고 입력하면 `input`에 해당하는 문자열은 `/roll 3d6` 전체가 되는 것입니까? 아니면 `3d6` 뒷 부분만 해당되는 것입니까?

myeesan replied:
@jwChung 저희가 결정해야 하는 부분이라고 생각해요. `/roll 3d6`으로 입력받고 다른 명령어를 처리하기 위한 설계도 저희가 고민해야 할 부분이라고 생각합니다 :)

jwChung replied:
@myeesan 아 그렇군요. 제가 구현한 PR #22 은  `<message>` 부분을 `roll 2d3`로 간주하고 있습니다.

younggi commented:
dice를 roll 하는 core 기능을 만든다고 생각을 해보면, object name이 DiceBot보다 조금 하위 level의 이름이면  더 좋을 것 같다는 생각이 들지만, 어떤것이 좋을 지는 생각이 잘 안나네요.

jwChung commented:
@younggi 

> 조금 하위 level의 이름

이라고 표현하신게 더 구체적인 이름을 의미하시는 건가요?

younggi commented:
slack의 api request를 처리하는 부분까지를 포함한 부분 DiceBot이라 생각이 들어서 의견 적어봤습니다. 
하위 level이란 이름은 모호했네요.. 
- request 처리 부분 => 정렬님 부분
- dice bot logic 부분 => 진욱님 부분 
  ==> 전체가 DiceBot 인가 하는 생각도 들어서...

작은 시작이라 큰 문제는 없겠지만, 전체 slack bot에서 layer가 구분되면 함께 작업하는데 좀 더 편할까..하는 생각에 적어봤습니다.

funbot 이니 다른 bot도 만들어 넣을 수 있으면 좋겠다는 생각도 들었네요...

younggi commented:
말씀하신 좀 더 구체적인 이름일 수도 있는데, 일단 그냥 진행하셔도 됩니다.~ ^^

jwChung commented:
@younggi 네 좋은 의견이십니다. 그런데 지금 상태에서는 디자인을 최대한 고려하지 않고 Dice Bot 구현에 대해서만 촛점을 맞추는게 어떨까요? 아마 @kwoolytech 님도 #14 에서 비슷한 고민을 하고 계시지 않나 생각됩니다.

b01001011 commented:
Decent start.

younggi commented:
넵~ 고고~ 하시죠^^

jwChung commented:
@younggi 넵 감사합니다. :) @myeesan 님 리뷰의견이 궁금한데 듣고 진행하는 건 어떨까요?

myeesan commented:
넵 방금 서버 테스트 완료했습니다. 우선 Request 부분 핸들링은 다음 오프 이후에 결정하는게 좋을 것 같습니다. 현재 Web hook 방식 외에 Command 방식도 있는데 어떤 방식을 사용할지 결정해야 하고요, Request를 직접 처리하려면 웹 프레임워크를 다뤄야 해서요. 우선은 dice bot 구현 부터 완료하고 진행해도 무리 없다고 생각합니다.

jwChung commented:
다음 시간에 [The Rules of the Open Road](http://blog.half-ogre.com/posts/software/rules-of-the-open-road/) 내용 같이 한번 봤으면 좋겠네요. 아래는 이 문서의 일부를 발췌한 것입니다.

> When you think the code is done, don’t merge it into master; instead, open a pull request from your working branch to master. This pull request will give your fellow contributors a place to review the code, and to discuss your work. Once the review is done and the discussion has ended, one of the reviewing contributors will merge the pull request. You should never merge your own pull request. Having another contributor merge the pull request does one better than just saying “Ship It!”.

몇몇 분들이 이 PR 승인 의견을 주셨습니다. merge해도 되겠다는 판단이드는데요. 그런데 제가 직접 merge를 하는 것은 위 내용과 같이 적절하지 않다고 생각되네요. 다른 분이 merge 해주시면 좋겠네요.

jwChung committed:
Add the document on the process method.

jwChung committed:
Refactored DiceBot from Object to Class,

I think that refactoring from Object to Class is more difficult than the reverse.

b01001011 commented on src/test/scala/DiceBotTest.scala:
```diff
@@ -0,0 +1,14 @@
+    test(s"process with invalid input '$x' returns none.") {
+      val expected = None
+      val actual = DiceBot.process("invalid string")
+      assert(expected == actual)
```
The way you write the test code is somewhat different with the test codes I've seen on the Exercism. What is advantage of using  `Seq ... foreach ...` thing? I have no issue with the way you writing it. Just want to know. :)

jwChung replied:
Hm. The idea is from [the parameterized test](http://xunitpatterns.com/Parameterized%20Test.html). Usually tests that you have encountered are [Tabular Test](http://xunitpatterns.com/Parameterized%20Test.html#Tabular%20Test) when you need tests but only test parameters are different from each others.

The following quotation sharply describes what is the different between the parameterized test and the tabular test.

> Several early reviewers wrote to me about a variation they use regularly: the Tabular Test. The essence of this is the same as doing a Parameterized Test except that the entire table of values is in a single Test Method. Unfortunately, this makes the test an Eager Test (see Assertion Roulette on page X) because it verifies many test conditions. This isn't a problem when all the tests are passing but it does lead to a lack of Defect Localization (see Goals of Test Automation) when one of the rows fails.

b01001011 commented on src/main/scala/DiceBot.scala:
```diff
@@ -1,3 +1,13 @@
+    *              respectively.
+    * @return returns `None` if the `input` is invalid, otherwise `Some[String]`
+    *         representing the order.
+    */
```
👍 good description. Maybe, we might need to set a ground rule for the comments as well.

jwChung replied:
Good point! We should discuss how to write documents.

> we might need to set a ground rule for the comments as well.

b01001011 commented:
So, your intention on this PR was to create a skeleton of the dice class and the test code. Good job on processing everything so far!  FYI, I am seeing this message, `Only those with write access to this repository can merge pull requests.`

b01001011 commented:
I have updated the decription so that people may know the scope of this PR. Hope you don't mind. :)

jwChung commented:
@kwoolytech 

> Only those with write access to this repository can merge pull requests.

I am going to address this issue. Thank you for pointing that out.

> I have updated the decription so that people may know the scope of this PR. Hope you don't mind.

I must say 👍 "히트다 히트" :)

jwChung committed:
Update document of process

The 'roll' command word would be considered.
